### PR TITLE
Simplify cancel command

### LINKED ISSUE #1530: Simplify cancel command
The cancel command can be slow for large builds, taking many hours to run. In previous tickets we have improved performance for the resume command, which is similar to cancel. We should use the same techniques to improve performance for the cancel command as well.

### PR DESCRIPTION
**Pull request by tst-lsavoie**
Updates the cancel command to use graph-based processing.

Testing indicates that this operation is about 25% slower than the legacy cancel operation. This is not terribly surprising because the graph code has to load everything from disk up front to create the graph, whereas the legacy code can load things as needed. On the flip side, the legacy code can end up loading the same asset multiple times whereas the graph code doesn't have that problem. The cancel operation happens to avoid the problem of loading the same asset multiple times reasonably well, so the legacy code tends to work better. Accordingly, I have updated the GraphOperations flag to allow for no graph operations, only graph operations that improve performance, or all graph operations. This allows users to use the new cancel if it is helpful to them or to avoid it and use the other graph operations to get the best possible performance. As we optimize the graph code more it will make more sense to use the cancel operation.

There are no release notes updates for this PR. Since it's not yet to a point that we think users should use it, I thought it best to leave it out of the release notes.

Fixes #1530.

### CODE REVIEW

File: earth_enterprise/src/fusion/autoingest/sysman/StateUpdater.cpp [hunk 1]
```diff
@@ -276,7 +279,9 @@ void StateUpdater::SetStateForRefAndDependents(
     function<bool(AssetDefs::State)> updateStatePredicate) {
   try {
     SharedString verref = AssetVersionImpl::Key(ref);
```
[the remark this answers is not in the document]
Comment (tst-lsavoie) [reply earth_enterprise/src/fusion/autoingest/sysman/StateUpdater.cpp]: I have to admit being a little confused about what exactly is going on here. As far as I can tell, `ref` can come in several formats: an asset, a version, or it can ask for the "current" or "lastgood" version of a specific asset. The `Key` function sort of normalizes it so that "verref" always refers to a specific version number. I think that ref and verref are the same in some cases, but not all.

As far as whether `Bind` loads the version into the cache, I'm not certain about that. The code gets hard to follow pretty quickly. In some cases it has to load the asset (not the version) into the cache so that it can figure out which version is current or lastgood. There may be a naming issue here, but ultimately it does what it says - it returns the normalized key that you can use to refer to a specific asset version.

File: earth_enterprise/src/fusion/autoingest/sysman/AssetOperation.cpp [hunk 2]
```diff
@@ -63,8 +63,38 @@ void RebuildVersion(const SharedString & ref, bool graphOps) {
   }
 }
 
-void HandleTaskProgress(const TaskProgressMsg & msg, bool graphOps) {
-  if (graphOps) {
+void CancelVersion(const SharedString & ref, MiscConfig::GraphOpsType graphOps) {
```
[the remark this answers is not in the document]
Comment (tst-lsavoie) [reply earth_enterprise/src/fusion/autoingest/sysman/AssetOperation.cpp]: Yeah, I always wonder whether passing a reference at all makes sense because SharedStrings are designed to be easy to copy. But hypothetically we could eventually change refs to something that doesn't copy easily, at which point references are more efficient. I figure const references aren't worse than anything else, so I just go with that by default.

File: earth_enterprise/src/common/khxml/khidl.pl [hunk 3]
```diff
@@ -1918,10 +1918,11 @@ sub EmitEnumDOMReader
     my $i = 0;
     for ($i = 0; $i < @{$enum->{enumerators}}; ++$i) {
 	my $item = $enum->{enumerators}[$i];
+    my $conditional = "enumStr == \"$item->{name}\" || enumStr == \"$item->{value}\"";
```
[the remark this answers is not in the document]
Comment (tst-lsavoie) [reply earth_enterprise/src/common/khxml/khidl.pl]: I don't see it. What's the issue?

[the remark this answers is not in the document]
Comment (tst-sbayard) [reply earth_enterprise/src/common/khxml/khidl.pl]: It looks like that line is indented a couple of spaces to the left of the line before and after it.

File: earth_enterprise/src/fusion/autoingest/sysman/StateUpdater.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (tst-sbayard) [reply earth_enterprise/src/fusion/autoingest/sysman/StateUpdater.cpp]: Ok, I didn't catch the normalization aspect of `Key`.

File: earth_enterprise/src/common/khxml/khidl.pl (hunk 3 above)
[the remark this answers is not in the document]
Comment (tst-lsavoie) [reply earth_enterprise/src/common/khxml/khidl.pl]: Got it. Fixed.
